V1 Unit Tests for Tags

andrewhand opened the pull request:
Added code for WIP tags unit tests.

jkglasbrenner commented:
Fixed the rebase (the merge diff was claiming more than one file was changed) and force-pushed, make sure to reset (or delete and recreate) your branches accordingly.

jkglasbrenner commented on tests/unit/restapi/test_tags_v1.py:
```diff
@@ -0,0 +1,525 @@
+    )
+
+
+def register_tag(
+    client: FlaskClient,
+    name: str,
+    group_id: int | None = None,
+) -> TestResponse:
+    """Register a tag using the API.
+
+    Args:
+        client: The Flask test client.
+        name: The name to assign to the new tag.
+        group_id: The group to create the new tag in.
+        description: The description of the new tag.
+
+    Returns:
+        The response from the API.
+    """
+
+    payload = {"name": name}
+
+    if group_id:
+        payload["group_id"] = group_id
```
`group_id` is not optional when creating a tag, so I suggest revising this as:

```suggestion
def register_tag(
    client: FlaskClient,
    name: str,
    group_id: int,
) -> TestResponse:
    """Register a tag using the API.

    Args:
        client: The Flask test client.
        name: The name to assign to the new tag.
        group_id: The group to create the new tag in.

    Returns:
        The response from the API.
    """

    payload = {"name": name, "group_id": group_id}
```

jkglasbrenner commented on tests/unit/restapi/test_tags_v1.py:
```diff
@@ -0,0 +1,525 @@
+
+    payload = {}
+
+    if Q:
+        payload.update(Q)
+    if G:
+        payload.update(G)
+    if P:
+        payload.update(P)
```
Query parameters and JSON payloads need to be separated out, see requested changes in #447 for more details.

jkglasbrenner reviewed (changes requested):
I left a couple of inline code comments, but the majority of changes I would request are the same as #447. This pull request should be updated in the same manner after #447 is resolved.

keithmanville commented on tests/unit/restapi/lib/actions.py:
```diff
@@ -101,3 +102,32 @@ def register_queue(
+def register_tag(
+    client: FlaskClient,
+    name: str,
+    group_id: int | None = None,
```
group_id should be be required. I checked and this is actually missing from the request schema. Can you add it there as well?

andrewhand replied:
Added with commit: https://github.com/usnistgov/dioptra/pull/455/commits/d89f7da3dbebebb37ff6f2a189405d92ae74d803

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    assert response.status_code == 200 and response.get_json()["data"] == expected
+
+
+def assert_registering_existing_tag_name_fails(client: FlaskClient, name: str) -> None:
```
group_id needed

andrewhand replied:
Added with commit: [fc5634a](https://github.com/usnistgov/dioptra/pull/455/commits/fc5634ad358575a409bf112ac194fc9cb5442fc7)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    Raises:
+        AssertionError: If the response status code is not 400.
+    """
+    response = actions.register_tag(client, name=name)
```
group_id needed

andrewhand replied:
Added with commit: [fc5634a](https://github.com/usnistgov/dioptra/pull/455/commits/fc5634ad358575a409bf112ac194fc9cb5442fc7)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    assert_tag_response_contents_matches_expectations(
+        response=tag_expected,
+        expected_contents={
+            "name": tag_expected["name"],
```
We don't want to use the output of registered tag here, since we want to ensure the response matches what the user input.

Instead create a variable `name="tag"` and pass it to both the register function and this assert.

andrewhand replied:
Added with commit: [58ffe1f](https://github.com/usnistgov/dioptra/pull/455/commits/58ffe1f8d4b851bc255d8a93e2608f3c331483c2)

keithmanville commented on tests/unit/restapi/v1/conftest.py:
```diff
@@ -90,3 +90,29 @@ def registered_queues(
+) -> dict[str, Any]:
+    tag1_response = actions.register_tag(
+        client,
+        name="ML: Object Recognition",
```
@jkglasbrenner Did we want to limit what characters can be used in tag names? I was thinking what I would typically think of as a tag would be something like `[a-zA-Z0-9_\-]`. I see it's just `Text` in the DB.

jkglasbrenner replied:
@keithmanville Since we're borrowing the tag concept from GitHub's labels, I just attempted to make a label using spaces and special characters, looks like it works:

![image](https://github.com/usnistgov/dioptra/assets/8547754/6845466c-506f-48e0-9c4e-ce3d0cb3630d)

So I'm okay with not being restrictive on the tag names. Users/Groups always have the option to set their own agreed-upon rules for how they will make tags.

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    """
+    existing_tag = registered_tags["tag1"].get_json()
+
+    assert_registering_existing_tag_name_fails(client, name=existing_tag["name"])
```
needs group_id

andrewhand replied:
Added with commit: [fc5634a](https://github.com/usnistgov/dioptra/pull/455/commits/fc5634ad358575a409bf112ac194fc9cb5442fc7)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+
+    assert_tag_name_matches_expected_name(
+        client, tag_id=tag1["tagId"], expected_name=tag1["name"]
+    )
```
Don't need this. Already testing creation of tags in another test.

andrewhand replied:
Removed assert in commit: [7ba6a3f](https://github.com/usnistgov/dioptra/pull/455/commits/7ba6a3f44ae742c02f3667c15e1e46fd196cb34b)

keithmanville commented on tests/unit/restapi/v1/conftest.py:
```diff
@@ -90,3 +90,29 @@ def registered_queues(
+        client,
+        name="ML: Object Recognition",
+        group_id=auth_account["default_group_id"],
+    )
```
fixtures should return `dict` not `TestResponse`. Add `.get_json()` to the end of each call to `register_tag`

andrewhand replied:
Added in commit: [7ace257](https://github.com/usnistgov/dioptra/pull/455/commits/7ace257fc286888bd374327230b9be74fa78ef47)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+      during registration.
+    """
+    tag1_expected = registered_tags["tag1"].get_json()
+    tag2_expected = registered_tags["tag2"].get_json()
```
tag fixtures should already be dictionaries. Can remove `get_json` here.

andrewhand replied:
Removed in commit: [7936fb3](https://github.com/usnistgov/dioptra/pull/455/commits/7936fb3592b533d749157a8175217da0179a81d8)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    - The user retrieves information about the same tag and it reflects the name
+      change.
+    """
+    tag1 = registered_tags["tag1"].get_json()
```
tags fixtures should already be dictionaries. Can remove `get_json` here.

andrewhand replied:
Removed in commit: [7936fb3](https://github.com/usnistgov/dioptra/pull/455/commits/7936fb3592b533d749157a8175217da0179a81d8)

keithmanville commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    assert_retrieving_tag_by_id_works(
+        client, tag_id=tag_expected["tagId"], expected=tag_expected
+    )
+
```
nitpick: remove whitespace between function calls.

andrewhand replied:
Removed in commit: [a88fc61](https://github.com/usnistgov/dioptra/pull/455/commits/a88fc612b90fdd9d87bbdfd2aca99214458207c2)

andrewhand commented:
@keithmanville If you have already done a bit of review you might have noticed a incorrect fields type in tags/schema. This has been fixed in the most recent commit [e30ed21](https://github.com/usnistgov/dioptra/pull/455/commits/e30ed2112a3cecef7f5a58ee1ce7391fb891814d)

jkglasbrenner commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    """
+    tag_expected = registered_tags["tag1"]
+
+    assert_retrieving_tag_by_id_works(
+        client, tag_id=tag_expected["tagId"], expected=tag_expected
+    )
+    delete_tag(client, tag_id=tag_expected["tagId"])
+
+    assert_tag_is_not_found(client, tag_id=tag_expected["tagId"])
```
Same as other comment, I think you access the tag identifier with `id` instead of `tagId`:

```suggestion
    assert_retrieving_tag_by_id_works(
        client, tag_id=tag_expected["id"], expected=tag_expected
    )
    delete_tag(client, tag_id=tag_expected["id"])

    assert_tag_is_not_found(client, tag_id=tag_expected["id"])
```

jkglasbrenner commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    This test validates the following sequence of actions:
+
+    - A user registers three tags: "tag1", "tag2", "tag3".
+    - The user is able to retrieve a list of all registered tags with ...
```
Documentation: The line looks unfinished, need to replace `...` with something specific.

andrewhand replied:
Fixed in commit: [d6b7e43](https://github.com/usnistgov/dioptra/pull/455/commits/d6b7e4340aac661585c95919bcb0095b00d53854)

jkglasbrenner commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+
+    This test validates the following sequence of actions:
+
+    - A user registers a tag named "old_name".
+    - The user is able to retrieve information about the "old_name" tag that
+      matches the information that was provided during registration.
+    - The user renames this same tag to "new_name".
+    - The user retrieves information about the same tag and it reflects the name
```
Documentation: I don't think the tag starts with the name `"old_name"` anymore, from what I can tell it's `"ML: Object Recognition"`.

andrewhand replied:
Fixed in commit: [d6b7e43](https://github.com/usnistgov/dioptra/pull/455/commits/d6b7e4340aac661585c95919bcb0095b00d53854)

andrewhand commented on tests/unit/restapi/v1/test_tags.py:
```diff
@@ -0,0 +1,413 @@
+    existing_tag = registered_tags["tag1"]
+
+    assert_registering_existing_tag_name_fails(
+        client, name=existing_tag["name"], group_id=existing_tag["group_id"]
```
Fixed in commit: [9f4ac72](https://github.com/usnistgov/dioptra/pull/455/commits/9f4ac72b1db971e437bc9d8d469a1ff5e99efc4e)

andrewhand commented:
> In addition to the inline comments, we also mentioned during this morning's session that we've added a description field for all of the resources in Dioptra, so that's something that can be incorporated in this pull request.

Added descriptions to groups testing in commit: [415ae6c](https://github.com/usnistgov/dioptra/pull/455/commits/415ae6c0716b032c1a0480134ed2003a714116d3)

Will need a review before merge @jkglasbrenner @keithmanville
